ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/generator/IdGeneratorBase.java

```diff
@@ -0,0 +1,149 @@
+        Preconditions.checkArgument(null != constraints && !constraints.isEmpty());
+        nonceGenerator.registerGlobalConstraints(constraints);
+    }
+
+    public synchronized void registerDomainSpecificConstraints(
+            final String domain,
+            final IdValidationConstraint... validationConstraints) {
+        registerDomainSpecificConstraints(domain, ImmutableList.copyOf(validationConstraints));
+    }
+
+    public synchronized void registerDomainSpecificConstraints(
+            final String domain,
+            final List<IdValidationConstraint> validationConstraints) {
+        Preconditions.checkArgument(null != validationConstraints && !validationConstraints.isEmpty());
+        nonceGenerator.registerDomainSpecificConstraints(domain, validationConstraints);
+    }
+
+    /**
+     * Generate id with given namespace
+     *
+     * @param namespace String namespace for ID to be generated
+     * @return Generated Id
+     */
+    public Id generate(final String namespace) {
+        val idInfo = nonceGenerator.generate(namespace);
+        return nonceGenerator.getIdFromIdInfo(idInfo, namespace);
+    }
+
+    public Id generate(final String namespace, final IdFormatter idFormatter) {
+        val idInfo = nonceGenerator.generate(namespace);
+        return nonceGenerator.getIdFromIdInfo(idInfo, namespace, idFormatter);
+    }
+
+    /**
+     * Generate id that matches all passed constraints.
+     * NOTE: There are performance implications for this.
+     * The evaluation of constraints will take its toll on id generation rates.
+     *
+     * @param namespace     String namespace
+     * @param domain     Domain for constraint selection
+     * @param skipGlobal Skip global constrains and use only passed ones
+     * @return ID if it could be generated
+     */
+    public Optional<Id> generateWithConstraints(final String namespace, final String domain, final boolean skipGlobal) {
+        Optional<IdInfo> idInfoOptional = nonceGenerator.generateWithConstraints(namespace, domain, skipGlobal);
+        return idInfoOptional.map(idInfo -> nonceGenerator.getIdFromIdInfo(idInfo, namespace));
+    }
+
+    public Optional<Id> generateWithConstraints(final String namespace,
+                                                final List<IdValidationConstraint> inConstraints,
+                                                final boolean skipGlobal) {
+        Optional<IdInfo> idInfoOptional = nonceGenerator.generateWithConstraints(namespace, inConstraints, skipGlobal);
+        return idInfoOptional.map(idInfo -> nonceGenerator.getIdFromIdInfo(idInfo, namespace));
+    }
+
+    public Optional<Id> generateWithConstraints(final IdGenerationRequest request) {
+        val idInfo = nonceGenerator.generateWithConstraints(request);
+        return idInfo.map(info -> (nonceGenerator.getIdFromIdInfo(info, request.getPrefix(), request.getIdFormatter())));
+    }
+
```

Comment:
The issue is this implementation leaves too much for the specialization to do. Essential stuff like constrain validations etc should all be done in the base class. The _only_ responsibility of any specialization of the base should be to generate the id and hand it over to the base. The retry, collision checks etc should all happen in the base. The only function that need to be implemented by the specialization should be to generate an id based on it's logic

Reply:
Every other method should be marked as final in base